### ISSUE TITLE
Offline mode release

### ISSUE THREAD

**Issue by jose-pr** [MicrosoftDocs/intellicode#102]
At the moment the extensions requires an Internet Connection to download the schema.
 This doesn't work for completely offline or intranet networks/devices. Can there be an offline installation package which comes with all the schemas and can be manually downloaded? Or an easy setting in VS Code to point to a folder to look for schemas for?

**Comment by jkeech** [comment]
IntelliCode does require an internet connection to acquire the model initially, but after the first download, it should work in a fully offline environment. At startup, the IntelliCode extension will always fall back to a locally cached model if one exists.

How do you normally acquire extensions and and updates today? We don't currently have any plans to support different mechanisms to download/acquire models.

**Comment by jose-pr** [comment]
There are systems which normally acquire the packaged by downloading the vsix , passing it through viru scan and then uploading it to one server on the intranet network with no access to the internet. The actual computer which uses it never touches the internet. So if the package came with them by default in case in cant access the internet it would work, as if it has internet it can get an updated copy if not used the one it has. Or if not making it easier to set them up that doesnt require creating directories and a custom config file maybe a setting with the path per schema or folder with all schemas. And then i can also download the schemas and push them up with the same process. 
Extensions / update / programs everything is acquired once individually scanned and put on a server that then computers can access the file.

**Comment by jkeech** [comment]
Thanks for the details. This is not a supported scenario for now, but if there is enough user demand for true offline support, than it's something we will investigate. Thanks!

**Comment by jose-pr** [comment]
Thanks for looking into it. Love VS Code but it makes it harder to use when it "doesn't work out of the box". Even VS Code tries to get schemas files when it opens for the first time which would be fine if it was only looking for updates but not missing them from the get go.
Related Issued:
 https://github.com/MicrosoftDocs/intellicode/issues/4#issuecomment-450935237 On how to do this manually and problems that occur due to proxy and could be alleviated by including an initial schema for in the package.
https://github.com/MicrosoftDocs/intellicode/issues/46 
https://github.com/MicrosoftDocs/intellicode/issues/99
Basically this feature is needed for highly protected networks with extremely limited access to the internet such as countries like China, corporate sensitive networks, or even goverment networks with limited to no access to the internet.

**Comment by iabbott** [comment]
We'd also like downloadable packages, much like the installable VSIX files... too many companies presume that everyone has direct Internet access, but there are many areas of software development that have to rely on centrally-controlled software distribution due to zero external network connectivity. In our case, every single VFX post-production house, due to MPAA security regulations has no direct access to the Internet (beyond read-only virtual terminals such as NoMachine). The majority of these houses develop in Python, so having access to these models would be very helpful, but with no directly downloadable package that can be transferred to an internal distribution mirror this new plugin is of no use whatsoever.

**Comment by syncush** [comment]
I would love to have an offline mode of intellicode.
In my workplace, we do not have direct internet access and every 2-3 weeks I can request packages and extensions for my IDE.
Surely you can now see why an offline mode is much needed.

**Comment by nirnir93** [comment]
Will be very useful in highly protected networks

**Comment by Shaharamir** [comment]
I will be happy if offline mode can be provided.
Thanks.

**Comment by orelkan** [comment]
I support this, I'd very much like an offline mode

**Comment by dormesica** [comment]
I also would like an offline mode. Mu company works on a closed network and it would help up very much

**Comment by santornl** [comment]
I would like an offline mode

**Comment by scandinave** [comment]
Me too

**Comment by JoniVirtanenWork** [comment]
This is definitely important feature (core feature, may I say). There are countless environments where network connectivity just isn't available.
At least there should be non-AI version, which relies only on latest loaded model.
In general making stuff network dependent makes them susceptible to privacy breaches, industrial espionage, extra costs from network operator and of course application or feature can at any time become available only if purchased.

**Comment by markw-t** [comment]
Thanks for the comments @JoniVirtanenWork - I think there might be a small misunderstanding regarding "non-AI version", so I wanted to clarify.  IntelliCode's completions are always based on a machine learning model, and once that model is loaded it can be used. After that first download, we don't require network connection in order to provide IntelliCode completion recommendations, and there is no ongoing learning process going on that requires a network connection.  We do need a network connection if you want to create a custom model for your own codebase rather than relying on the base model which is trained on thousands of open source repos. I hope that makes sense?

Having said all of that, we recognize that IntelliCode won't currently work if your development machine is _never_ connected to the internet, and that is what this feature request is about.

**Comment by COD3BREAK3R** [comment]
Could your online dependencies be added within the same extension to work 100% Offline? I think that is not very difficult to do, right?

**Comment by mikesmith1611** [comment]
This would be awesome

**Comment by JoniVirtanenWork** [comment]
Thanks for the clarification @markw-t. I have an environment, where all data and files are copied manually from extranet machine. Can I just copy certain related files from connected machine (which has done the model downloading) over to offline machine and have working intellisense?
If manual copying works, which files should I copy?

**Comment by ChristianHaefner** [comment]
Offline mode would be great!

**Comment by ChubbyArse** [comment]
I too woud welcome an offline mode.

**Comment by RejectKid** [comment]
I would like this

**Comment by jamesboyer** [comment]
There is no way to use Intellicode without it supporting offline for me, so I would also love this feature

**Comment by J-Sorenson** [comment]
First off, at a minimum, any extension that relies on this (Python) should state that it is not a fully-offline installer.  The network requirement is not even listed on the Python Extension documentation at the Marketplace or GitHub page.

We recently installed VS Code and  CPPTools extension on our secure network at Lockheed Space Company. Since we had the Anaconda Python distribution installed, I requested importing the Python plugin to our network.  Ouch! The undocumented limitation of the offline installer was a bit embarrassing for me.  At least the Jedi/PyLint fallback makes it somewhat usable, but it's very apparent that most of the recent effort is devoted to the language server that we have no capability of installing.

In this day and age of cyber-security, you'll find that more companies are sealing off their networks from the internet,  We have machines that are connected to the "outside" world, but that is for communication and references.  All imports to the secured network(where work is done) are requested, scanned, transferred, then installed to a sandbox.  I'm a little surprised that an offline installer was not an assumed feature.  VS Code and most of it's extensions already support it.  This is an outlier for a rather critical future component.

Long-story short: of course I'm going to up-vote this.  It's a requirement for any software that we use. I suppose I can ask some of our python engineers to up-vote this as well, if that helps with the decision.

**Comment by J-Sorenson** [comment]
There is a download link to a .visx at https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode, and the internal files are dated from last August.  Can that be configured and used as an offline installer, or will it not work with the Python extension?  I'd like to know prior to having the admins go through security checks (already got bit once).

**Comment by mirekphd** [comment]
> There is a download link to a .visx at https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode, 

It's not a link, but a button, and the link is hidden in source code. They go to great lengths to hide the source from this allegedly open source piece of software...

**Comment by J-Sorenson** [comment]
The plugins are not necessarily required to be open-source (per the license), so that's not my concern right now.  However, I DO require an actual offline installer for this to be usable.  It sounds like my only only option is to somehow extract the files from a downloaded installation on the external network, but we don't have any Linux servers on the outside. Besides, that may raise a few eyebrows with the network security team as I request to pull the files over.

I'll try looking into this again sometime in December.  Until then, I guess we'll stick with PyCharm for now.  Honestly, I'd even look into a Microsoft Visual Studio license if they made a Linux version.

**Comment by mirekphd** [comment]
At least the direct download link to any VS Code extension can be obtained even without any releases available on Github (which would be the standard installation method used in most docker builds) using VS Code _vsassets.io_ API:

PKG_PUBLISHER=VisualStudioExptTeam
PKG_NAME=vscodeintellicode
PKG_VER=1.2.0

https://ms-vscode.gallery.vsassets.io/_apis/public/gallery/publisher/VisualStudioExptTeam/extension/vscodeintellicode/1.2.0/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

The publisher and package name can be obtained from the Visual Studio Marketplace URL of your chosen package, e.g.:
https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode
while the latest version, from the Changelog linked from on the same page, here::
https://marketplace.visualstudio.com/items/VisualStudioExptTeam.vscodeintellicode/changelog

As for the model object being pulled from the net at first run, it turns out that there is a less hacky method than an unrestricted sandbox server outside of the corporate network (which is a must for air-gapped servers in my experience...), because VS has some shared models import tool, so maybe this is be available in VS code as well (the binary file needs to be downloaded anyway, so I myself will try bundling it with the container repo if size permits):
https://docs.microsoft.com/en-us/visualstudio/intellicode/share-models#add-a-user-associated-team-completions-model

**Comment by mirekphd** [comment]
> may raise a few eyebrows with the network security team as I request to pull the files over.

As a side note: you may rethink your setup, because there is no need to clear every constituent file and having cleared all be permitted to build the container. There are excellent tools for scanning entire docker containers for vulnerabilities. One of them is Clair, a long-standing open source security scanner, part of the recently open-sourced Quay. I use it with an open source client and report generator named Yair.

**Comment by mirekphd** [comment]
As for the model object being pulled from the net at first run, it turns out that there is a less hacky method than an unrestricted sandbox server outside of the corporate network (which is a must for air-gapped servers in my experience...), because both classic (desktop) VS and cloud-native VS Code (cdr/code-server) have models import tools.

After you install the vscodeintellicode extension in VS Code e.g. from the above link, go to Manage->Settings->Extensions->Visual Studio Intellicode to specify the model path (by default the extensions install folder):

```
Vsintellicode: Model Download Path
Specifies the folder path where downloaded model files are stored. Defaults to the extension install directory if not specified.
```

As for the classic desktop version of VS there are even more options:
https://docs.microsoft.com/en-us/visualstudio/intellicode/share-models#add-a-user-associated-team-completions-model

**Comment by OmniZenRaj** [comment]
Thanks for the visx link. I have downloaded & will see if I can set it up
to work completely offline. The usual problem is that when an extensions is
first run (not first installed), it will attempt to download some assets it
requires. This of course cannot be done offline, so I try to get the assets
from a connected online system and copy them to a local / network folder
for offline access. This doesn't usually work with many extensions
including this one. I will try to see if I can set it up and let you know.
Thanks again.

**Comment by zhaoshouren** [comment]
the lack of a full offline install, will prevent many government entities, such as federal agencies, from adopting use of this plugin.

**Comment by COD3BREAK3R** [comment]
Can someone explain me where can I download the files what need the extension to run offline and where I need copy the files for work the extension? Thanks

**Comment by ConceptJunkie** [comment]
Any news on offline installation capabilities for the the language models?   I started using VSCode, and am very impressed with it, but there are a few glaring omissions (e.g., visual merge in the diff tool, and the inability to do an offline installation of language models for Intellisense) that keep it from being the world-class tool it's clearly got the potential for.  

At work, our development VMs, and well, pretty much everything else minus our actual laptops, are severely restricted for Internet access.  e.g., the company maintains its own repos for RHEL and PyPI.   Installing VSIX files isn't a big deal, but there are no VSIX for the language models for Intellisense. 

Come on, guys.  VSCode is making me start to like Microsoft again, but help a brother out.

**Comment by yang-zhao** [comment]
I prompt this. Offline stuff must be essential as to product release and how could it ever been a disposal?

**Comment by J4s0nHu4n9** [comment]
Need offline support in my working environment too.

**Comment by sagivharel** [comment]
Offline support is a must for me

**Comment by mallorygiles** [comment]
I would also benefit from being able to at least be able to download the model and transfer that model to an offline machine to be able to use this extension

**Comment by tamnm** [comment]
We also need offline support.

**Comment by mrolinek** [comment]
Yes, please!

**Comment by Xiaoxiong-Liu** [comment]
Any update for offline support now?

**Comment by phydev** [comment]
Everyone working in a server for sensitive data under GDPR strict regulation, i.e. not internet access, will need offline support.  Please make this available.

**Comment by markw-t** [comment]
To ensure it's clear: we are now using this issue to track the entire notion of having IntelliCode's base completions feature work for machines that are fully offline with no internet access. It's what you should upvote if this scenario is important to you (as it is to many who have already commented and voted above)

**Comment by rlgloden** [comment]
Another vote for air-gapped installs.

**Comment by AntonyM71** [comment]
And another +1 for air gapped installs

**Comment by oersoy1** [comment]
> To ensure it's clear: we are now using this issue to track the entire notion of having IntelliCode's base completions feature work for machines that are fully offline with no internet access. It's what you should upvote if this scenario is important to you (as it is to many who have already commented and voted above)

Hi, is there any estimate for completion? Pardon me for being pushy but seems like many people need this and it has been in the radar for over 2 years now.  For the record, my team and I were asked to switch to servers that are completely off the internet so that we can work on sensitive data this we cannot use `Intellicode`

**Comment by max-sixty** [comment]
Has anyone tried the approach described by @vivlimmsft [here](https://github.com/MicrosoftDocs/intellicode/issues/4#issuecomment-450935237) from 2019 recently — that looks very promising, though potentially would need some tweaks for new versions?

Separately — if you want to leave a +1, then add a thumbs up to the post; there's really no need to create lots of notifications by announcing your +1 or asking for an update.

**Comment by MetricMike** [comment]
vivlimmsft's method works, *except* that `IntelliCodeModels-v2.json` is now `cache/models.json` relative to the extension folder (so `extensions/visualstudioexptteam.vscodeintellicode-1.2.14/cache/models.json`. Actual models can be anywhere, just be sure to update the `filePath` key with their location.

**Comment by max-sixty** [comment]
Thanks for replying @MetricMike .

Unfortunately I get errors (slightly abbreviated output):
```
Acquiring model `intellisense-members-lstm-pylance`
vs-intellicode-python was passed a model: {}.
Could not reach the IntelliCode service. Checking the local model cache.
None of the cached models match
None of the local models match.
No model v2 available for Python, trying v1.

Acquiring model `intellisense-members-lstm`
vs-intellicode-python was passed a model: {}.
Could not reach the IntelliCode service. Checking the local model cache.
None of the cached models match
None of the local models match.
No model v2 available for Python, trying v1.
```

@vivlimmsft 's solution has "analyzerName": "intellisense-members",`, rather than `intellisense-members-lstm` / `intellisense-members-lstm-pylance`, so maybe that's out of date? I tried changing the string in `models.json` but no luck. 

Is there a location we can download the later models?

**Comment by MetricMike** [comment]
@max-sixty try with the model at https://prod.intellicode.vsengsaas.visualstudio.com/api/v1/model/common/python/intellisense-members-lstm-pylance/output/latest ? Here's my models.json with that model, which appears to be working on my local? I don't work with python much so ¯\_(ツ)_/¯

I just switched out `intellisense-members` on the orignal post with `intellisense-members-lstm-pylance`. I suspect it's related to https://devblogs.microsoft.com/python/announcing-pylance-fast-feature-rich-language-support-for-python-in-visual-studio-code/ switching out the default language server last year and https://github.com/microsoft/vscode-python/issues/16204 that describes several detection issues around that switch.

**Comment by max-sixty** [comment]
I think I got it working, it was a fairly obvious replace. Here's a script that I think should package it up from a local machine, but YMMV, no assurances.


```bash
# Output of code command looks like
# `VisualStudioExptTeam.vscodeintellicode@1.2.14`
INTELLICODE_VERSION=$(code --user-data-dir=data/ --list-extensions --show-versions | rg -o 'intelli.*@(.*)' -r '$1')
INTELLICODE_PATH=data/extensions/visualstudioexptteam.vscodeintellicode-${INTELLICODE_VERSION}
# Model URL is something like
# https://intellicodeprodstorage.blob.core.windows.net/models-prod-output/Output_88630A1B4C077752096ECAA576B92C7D3FA1_9BA695AD4C514C32A5D0AC9B02846CD7?skoid=511b33f5-16ee-4db4-a582-21e2be0519ae&sktid=975f013f-7f24-47e8-a7d3-abc4752bf346&skt=2021-07-22T06%3A24%3A16Z&ske=2021-07-23T06%3A24%3A16Z&sks=b&skv=2020-02-10&sv=2020-02-10&se=2021-07-22T07%3A34%3A59Z&sr=b&sp=r&sig=XI3m1%2BIHKIrr6yB8EciA3UZ7Xzjwd7GhqYjiHCsRXrc%3D
MODEL_URL=$(curl https://prod.intellicode.vsengsaas.visualstudio.com/api/v1/model/common/python/intellisense-members-lstm-pylance/output/latest | jq -r '.. | .readSasToken? //empty')
MODEL_NAME=$(echo $MODEL_URL | rg -o 'Output_(.*)\?' -r '$1')
MODEL_ID=$(echo $MODEL_NAME | rg -o '(.*)_' -r '$1')
MODEL_PATH=$INTELLICODE_PATH/cache/$MODEL_NAME

mkdir -p $INTELLICODE_PATH/cache
wget $MODEL_URL -q -O $MODEL_PATH

OFFLINE_MODEL_PATH=$$HOME/vscode/$MODEL_PATH

cat > $INTELLICODE_PATH/cache/models.json <<EOL
[
  {
    "analyzerName": "intellisense-members-lstm-pylance",
    "languageName": "python",
    "identity": {
      "modelId": "${MODEL_ID}",
      "outputId": "manually-downloaded",
      "modifiedTimeUtc": "2018-12-11T20:45:03.430Z"
    },
    "filePath": "${OFFLINE_MODEL_PATH}",
    "lastAccessTimeUtc": "2019-01-02T17:09:02.863Z"
  }
]

EOL
```

**Comment by max-sixty** [comment]
@MetricMike thanks a lot! I managed to miss your message. You were absolutely right. Hopefully both of ours are helpful for people

**Comment by ayim** [comment]
Another issue:
Language: Java
Connectivity: Offline server (Red Hat Enterprise Linux 7)
Other Extensions: Remote-SSH

https://github.com/MicrosoftDocs/intellicode/issues/238

**Comment by ayim** [comment]
@DavidObando I'm looking for engineering input here on...

the feasibility of packaging models in the extension
the current user experience of configuring a proxy in VSCode - I see a number of issues in the repo relating to the difficulty of configuring a proxy?

**Comment by ayim** [comment]
A few more upvotes here:
https://github.com/MicrosoftDocs/intellicode/issues/165

**Comment by ayim** [comment]
Another user with an unreliable internet connection:
https://github.com/MicrosoftDocs/intellicode/issues/233

**Comment by ayim** [comment]
Chinese users are unable to download the model, but seem to be able to download the extension:
https://github.com/MicrosoftDocs/intellicode/issues/99

**Comment by zivsivron** [comment]
is there any chance to get the model lib for java developers? im working in an air gapped enviroment for a company in israel where we use a CDR machine to wash files before installing on company computers. so i downloaded the extension and washed it and installed it with no problems, but once i start VScode , it tries to go online and download a model for intellicode which it cant. i have it installed on a private machine so maybe some1 can point me to the location of the models in a windows 10 system so i can copy and wash em and try to install em manually? thanks in advance.

**Comment by axel37** [comment]
Almost 4 years and this is still open :(

**Comment by birdhackor** [comment]
Does any one know how to disable model download and use model cache directly?

**Comment by vivlimmsft** [comment]
Last month we [released v1.3.0 to the marketplace](https://marketplace.visualstudio.com/items?itemName=VisualStudioExptTeam.vscodeintellicode) which no longer downloads models from a service, they are all included in the extension package now.
